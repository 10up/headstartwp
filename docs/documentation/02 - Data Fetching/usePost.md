---
slug: /data-fetching/usepost
sidebar_position: 1
---

# The usePost hook

> The [usePost](/api/modules/headstartwp_next#usepost) hook is the Next.js binding for the [useFetchPost](/api/namespaces/headstartwp_core.react#usefetchpost).

The `usePost` hook fetches a single WordPress post from a registered post type. Its basic usage is very simple.

## Basic Usage

The basic usage is pretty simple and it assumes a route named `src/pages/[...path].js` where the `slug` is extracted from the URL.

```js title=src/pages/[...path].js
import { usePost } from '@headstartwp/next';

const PostPage = () => {
	const { loading, error, data } = usePost();

	if (loading) {
		return 'Loading...';
	}

	if (error) {
		return 'error...';
	}

	return (
		<div>
			<h2>{data?.post.title.rendered}</h2>
		</div>
	);
};
```

#### Post path matching

:::caution
The behavior described here was implemented in version **0.5.x** of the framework and can be disabled by passing `matchCurrentPath: false` to `usePost` (and `fetchHookData` for server-side data fetching).
:::caution

The `usePost` hook will by default match the current path captured by `[...path].js` with the post's link property. This ensures the right post is loaded and that 404 are issued to unsupported permalinks. 

:::caution
The framework requires that the permalink structure set on the backend matches the URL structure being used on the front-end. 

This means that if you include the date in your permalinks the URLs to a post without the date will 404 in the framework.

This is also true for custom post types, if the WordPress backend expects a custom post type to be at `/book/book-name` the front-end must also follow the same URL structure. If you wish to change the permalink structure for custom post types make sure to make the changes both in WordPress and in your front-end code.
:::caution

Example where path matches:

- User visits URL `/post-name`
- The post with the `post-name` slug contains a `http://backend.com/post-name` link.
- Since the URL and the path of `post.link` matches the page/post is rendered.

Example where path does not match:

- User visits URL `/parent-page/post-name`
- The post with the `post-name` slug contains a `http://backend.com/different-parent/post-name` url
- Since the URL (`/parent-page/post-name`) and the path of `post.link` (`different-parent/post-name`) **do not** match a 404 page is issued.

Example where path does not match but is redirected to the right one:
- User visits URL `/post-name`
- The post with the `post-name` slug contains a `http://backend.com/2022/10/30/post-name` url
- Since the URL and the path of `post.link` do not match, a NotFound error is thrown
- If prefetching is setup following [pre-fetching](/learn/data-fetching/prefetching) and `redirectStrategy` is set to "404" or "always" in `headless.config.js`, `handleError` will then look if there's a redirect available and since WordPress redirects `/post-name` to `/2022/10/30/post-name`, the framework will also perform the redirect.


### Fetching from multiple post types

When specifying an array of post type, the slug will be searched in both endpoint and the first one to return a valid post object will be used.

:::caution
This might result in URL conflicts, i.e a post or page using the same slug. The first post type specified will take precedence. In such cases, consider using a different URL structure for each for instance (e.g: using `src/pages/article/[...path].js` for posts).
:::caution


```js title="src/pages/[...path].js"
import { usePost } from '@headstartwp/next';

const PostOrPage = () => {
	const { loading, error, data } = usePost({ postType: ['page', 'post'] });

	if (loading) {
		return 'Loading...';
	}

	if (error) {
		return 'error...';
	}

	return (
		<div>
			<h2>{data?.post.title.rendered}</h2>
		</div>
	);
};
```

## Fetching from a custom post type

To fetch a single from a custom post type, first declare the custom post type in `headless.config.js` as explained in the [headless.config.js](/learn/getting-started/headless-config#custom-post-types) section.

```js title="src/pages/book/[...path].js"
import { usePost } from '@headstartwp/next';

const PostPage = () => {
	const { loading, error, data } = usePost({ postType: 'book' });

	if (loading) {
		return 'Loading...';
	}

	if (error) {
		return 'error...';
	}

	return (
		<div>
			<h2>{data?.post.title.rendered}</h2>
		</div>
	);
};
```