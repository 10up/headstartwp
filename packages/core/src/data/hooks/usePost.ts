import { EndpointParams } from './types';
import { useFetch } from './useFetch';

const endpoint = '/wp-json/wp/v2/posts';

interface PostParams extends EndpointParams {
	slug?: string;
}

export function usePost(params: PostParams) {
	const { data, error } = useFetch(endpoint, params);

	if (!data || error) {
		return { data, error };
	}

	return { data: data[0] };
}
