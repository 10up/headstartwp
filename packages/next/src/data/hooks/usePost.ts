import {
	FetchResponse,
	getPostAuthor,
	getPostTerms,
	getWPUrl,
	PostEntity,
	PostParams,
	SinglePostFetchStrategy,
} from '@10up/headless-core';
import { SWRConfiguration } from 'swr';
import { useFetch } from './useFetch';
import { HookResponse } from './types';

export interface usePostResponse extends HookResponse {
	data?: { post: PostEntity };
}

/**
 * The usePost hook. Returns a single post entity
 *
 * @param params - Supported params
 *
 * @returns
 */
export function usePost(
	params: PostParams,
	options: SWRConfiguration<FetchResponse<PostEntity>> = {},
): usePostResponse {
	const { data, error } = useFetch<PostEntity, PostParams>(
		{ _embed: true, ...params },
		usePost.fetcher(),
		options,
	);

	if (error) {
		return { error, loading: false };
	}

	if (!data) {
		return { loading: true };
	}

	// TODO: fix types
	const post = Array.isArray(data.result) ? (data.result[0] as PostEntity) : data.result;

	post.author = getPostAuthor(post);
	post.terms = getPostTerms(post);

	return { data: { post }, loading: false };
}

usePost.fetcher = () => new SinglePostFetchStrategy(getWPUrl());
