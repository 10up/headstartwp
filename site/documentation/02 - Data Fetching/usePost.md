---
slug: /data-fetching/usepost
sidebar_position: 1
---

# The usePost hook

> The [usePost](/api/modules/10up_headless_next#usepost) hook is the Next.js binding for the [useFetchPost](/api/namespaces/10up_headless_core.react#usefetchpost).

The `usePost` hook fetches a single WordPress post from a registered post type. It's basic usage is very simple.

## Basic Usage

The basic usage is pretty simple and it assumes a route named `src/pages/[...path].js` where the `slug` is extracted from the URL.

```js
//src/pages/[...path].js
import { usePost } from '@10up/headless-next';

const PostPage = () => {
	const { loading, error, data } = usePost();

	if (loading) {
		return 'Loading...';
	}

	if (error) {
		return 'error...';
	}

	return (
		<div>
			<h2>{data?.post.title.rendered}</h2>
		</div>
	);
};
```

#### Post path matching

> The behavior described here was implemented in version **0.5.x** of the framework.

The `usePost` hook will by default match the current path captured by `[...path].js` with the post's link property. This ensures the right post is loaded and that 404 are issued to unsuported permalinks. 

> The scenarios described below assumes [pre-fetching](/docs/data-fetching/prefetching) is being used in Next.js

Example where path matches:

- User visits URL `/post-name`
- The post with the `post-name` slug contains a `http://backend.com/post-name` link.
- Since the URL and the path of `post.link` matches the page/post is rendered.

Example where path does not match:

- User visits URL `/parent-page/post-name`
- The post with the `post-name` slug contains a `http://backend.com/different-parent/post-name` url
- Since the URL (`/parent-page/post-name`) and the path of `post.link` (`different-parent/post-name`) **do not** match a 404 page is issued.

Example where path does not match but is redirected to the right one:
- User visits URL `/post-name`
- The post with the `post-name` slug contains a `http://backend.com/2022/10/30/post-name` url
- Since the URL and the path of `post.link` do not match, a NotFound error is thrown
- If prefetching is setup following [pre-fetching](/docs/data-fetching/prefetching) and `redirectStrategy` is set to "404" or "always" in `headless.config.js`, `handleError` will then look if there's a redirect avaliable and since WordPress redirects `/post-name` to `/2022/10/30/post-name`, the framework will also perform the redirect.


### Fetching from multiple post types

When specifying an array of post type, the slug will be searched in both endpoint and the first one to return a valid post object will be used.

> Note: this might result in URL conflicts, i.e a post or page using the same slug. The first post type specified will take precedence. In such cases, consider using a different URL structure for each for instance (e.g: using `src/pages/article/[...path].js` for posts).


```js
//src/pages/[...path].js
import { usePost } from '@10up/headless-next';

const PostOrPage = () => {
	const { loading, error, data } = usePost({ postType: ['page', 'post'] });

	if (loading) {
		return 'Loading...';
	}

	if (error) {
		return 'error...';
	}

	return (
		<div>
			<h2>{data?.post.title.rendered}</h2>
		</div>
	);
};
```

## Fetching from a custom post type

In order to fetch a single from a custom post type, first declare the custom post type in `headless.config.js` as explained in the [headless.config.js](/docs/getting-started/headless-config#custom-post-types) section.

```js
//src/pages/book/[...path].js
import { usePost } from '@10up/headless-next';

const PostPage = () => {
	const { loading, error, data } = usePost({ postType: 'book' });

	if (loading) {
		return 'Loading...';
	}

	if (error) {
		return 'error...';
	}

	return (
		<div>
			<h2>{data?.post.title.rendered}</h2>
		</div>
	);
};
```