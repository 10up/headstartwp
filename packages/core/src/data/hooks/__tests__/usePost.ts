import { renderHook } from '@testing-library/react-hooks';

import { usePost } from '../usePost';

jest.mock('next/router', () => ({
	useRouter() {
		return {
			route: '/',
			pathname: '',
			query: {},
			asPath: '',
		};
	},
}));

describe('usePost', () => {
	it('fetches a post', async () => {
		const { result, waitForNextUpdate } = renderHook(() => usePost({ slug: 'test' }));
		expect(result.current.data).toBeUndefined();
		await waitForNextUpdate();
		expect(result.current.data).toMatchSnapshot();
	});

	it('fails on a 404 post', async () => {});
});
