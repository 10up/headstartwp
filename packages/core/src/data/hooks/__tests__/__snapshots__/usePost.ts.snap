// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`usePost fetches a post 1`] = `
Object {
  "_links": Object {
    "about": Array [
      Object {
        "href": "https://js1.10up.com/wp-json/wp/v2/types/post",
      },
    ],
    "author": Array [
      Object {
        "embeddable": true,
        "href": "https://js1.10up.com/wp-json/wp/v2/users/3",
      },
    ],
    "collection": Array [
      Object {
        "href": "https://js1.10up.com/wp-json/wp/v2/posts",
      },
    ],
    "curies": Array [
      Object {
        "href": "https://api.w.org/{rel}",
        "name": "wp",
        "templated": true,
      },
    ],
    "replies": Array [
      Object {
        "embeddable": true,
        "href": "https://js1.10up.com/wp-json/wp/v2/comments?post=57",
      },
    ],
    "self": Array [
      Object {
        "href": "https://js1.10up.com/wp-json/wp/v2/posts/57",
      },
    ],
    "version-history": Array [
      Object {
        "count": 0,
        "href": "https://js1.10up.com/wp-json/wp/v2/posts/57/revisions",
      },
    ],
    "wp:attachment": Array [
      Object {
        "href": "https://js1.10up.com/wp-json/wp/v2/media?parent=57",
      },
    ],
    "wp:term": Array [
      Object {
        "embeddable": true,
        "href": "https://js1.10up.com/wp-json/wp/v2/categories?post=57",
        "taxonomy": "category",
      },
      Object {
        "embeddable": true,
        "href": "https://js1.10up.com/wp-json/wp/v2/tags?post=57",
        "taxonomy": "post_tag",
      },
    ],
  },
  "author": 3,
  "categories": Array [
    1,
  ],
  "comment_status": "open",
  "content": Object {
    "protected": false,
    "rendered": "<h3>Fugit nemo placeat eos aliquam repellendus. Qui maxime quam hic sit</h3>
<p><img class=\\"alignleft\\" alt=\\"Autem ea laborum iste adipisci sed maxime animi est labore sit\\" src=\\"https://js1.10up.com/wp-content/uploads/2020/05/88670812-1b59-3095-9c92-8ae541f3b6d2.jpg\\"></p>
<h2>Maxime atque animi et nihil et. Distinctio adipisci doloremque placeat consectetur aut blanditiis sint</h2>
<p><!--more--></p>
<blockquote><p>Totam eos eos eligendi voluptas voluptates eaque. autem magnam quam voluptatem autem. Voluptatem consectetur <a title=\\"At facilis eaque.\\" href=\\"https://www.swift.com/illo-illum-deleniti-et-quia\\">accusamus aspernatur</a> voluptates facilis. Voluptas <a title=\\"Consequuntur eligendi.\\" href=\\"http://www.littel.com/minima-voluptas-eos-magnam-dolores-porro-sunt.html\\">deleniti aperiam aut omnis</a> Consectetur officiis cupiditate voluptates <a title=\\"Hic.\\" href=\\"http://pagac.com/perspiciatis-qui-aut-ad-iste\\">iure maiores. Debitis voluptatem</a> optio omnis. Officia excepturi eum <a title=\\"Quia sint eaque.\\" href=\\"http://www.kutch.org/\\">molestiae</a> aut. Saepe ipsum reprehenderit qui est. Quod placeat ducimus reprehenderit dolores maiores placeat architecto. Voluptates et <a title=\\"Deleniti ullam modi.\\" href=\\"http://www.runte.com/fuga-nostrum-totam-voluptas-fuga-hic-sunt\\"><a title=\\"Nihil laboriosam officiis optio.\\" href=\\"http://www.hammes.com/non-qui-dolore-sit-tenetur-veritatis-id.html\\">qui. Deleniti id id aperiam necessitatibus recusandae.</a></a> velit voluptatum nesciunt. ut porro minus dolorem sint ullam. aliquid quia Eaque id natus sequi consectetur <a title=\\"Aut ut eum molestiae optio sint omnis ratione.\\" href=\\"http://www.conroy.info/sapiente-placeat-est-voluptatum-delectus\\">est voluptatem. Nisi voluptas dicta impedit. Deleniti</a> tenetur molestias architecto rem magni <a title=\\"Quia id voluptatem corporis.\\" href=\\"https://www.connelly.com/nihil-animi-nesciunt-vero-sint-et-est\\">Sequi deserunt animi aut</a> quia doloribus enim. Aut omnis facilis odio ad. Quidem est repellat. optio nihil sed.</p></blockquote>
<p><img class=\\"alignleft\\" alt=\\"Qui omnis iste\\" src=\\"https://js1.10up.com/wp-content/uploads/2020/05/348bed06-c351-381e-8ed9-1587fcc172d8.png\\"></p>
<ul>
<li>Aut enim asperiores sed dicta</li>
<li>Maiores</li>
</ul>
<h3>Sit quia velit voluptas. Aliquid deserunt qui aspernatur soluta ullam quidem. Ab nihil quasi optio ut ex</h3>
<p><img alt=\\"Minima\\" src=\\"https://js1.10up.com/wp-content/uploads/2020/05/41dfb8fb-ae61-390a-b53f-1f98e23f462f.jpg\\"></p>
<hr>
",
  },
  "date": "2020-05-07T23:12:51",
  "date_gmt": "2020-05-07T23:12:51",
  "excerpt": Object {
    "protected": false,
    "rendered": "<p>Fugit nemo placeat eos aliquam repellendus. Qui maxime quam hic sit Maxime atque animi et nihil et. Distinctio adipisci doloremque placeat consectetur aut blanditiis sint</p>
",
  },
  "featured_media": 0,
  "format": "standard",
  "guid": Object {
    "rendered": "https://js1.10up.com/2020/05/07/modi-qui-dignissimos-sed-assumenda-sint-iusto/",
  },
  "id": 57,
  "link": "https://js1.10up.com/2020/05/07/modi-qui-dignissimos-sed-assumenda-sint-iusto/",
  "meta": Array [],
  "modified": "2020-05-07T23:12:51",
  "modified_gmt": "2020-05-07T23:12:51",
  "ping_status": "open",
  "slug": "modi-qui-dignissimos-sed-assumenda-sint-iusto",
  "status": "publish",
  "sticky": false,
  "tags": Array [],
  "template": "",
  "title": Object {
    "rendered": "Modi qui dignissimos sed assumenda sint iusto",
  },
  "type": "post",
}
`;
